### PR TITLE
[llvm][SelectionDAG] Fix up chains in lowerInvokeable. rdar://113994760

### LINKED ISSUE #64826: LLVM ERROR: sleb128 and uleb128 expressions must be absolute
```
$ cat <<EOF > rdar113994760.ll
source_filename = "rdar113994760.ll"
target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-apple-macosx"

define swiftcc void @"$s4main11call_foobaryyF"() personality ptr @__gcc_personality_v0 {
entry:
  %swifterror = alloca swifterror ptr, align 8
  invoke swiftcc void null(i64 0, ptr null, ptr swifterror %swifterror)
          to label %.noexc unwind label %tsan_cleanup

.noexc:                                           ; preds = %entry
  ret void

tsan_cleanup:                                     ; preds = %entry
  %cleanup.lpad = landingpad { ptr, i32 }
          cleanup
  resume { ptr, i32 } zeroinitializer
}

declare i32 @__gcc_personality_v0(...)

EOF
```

```
$ ./build/bin/llc rdar113994760.ll -filetype=obj -o out
LLVM ERROR: sleb128 and uleb128 expressions must be absolute
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: ./build/bin/llc rdar113994760.ll -filetype=obj -o out
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  llc                      0x0000000105f11d60 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  llc                      0x0000000105f12134 SignalHandler(int) + 188
2  libsystem_platform.dylib 0x00000001a87e1a84 _sigtramp + 56
3  libsystem_pthread.dylib  0x00000001a87b2c28 pthread_kill + 288
4  libsystem_c.dylib        0x00000001a86c1b28 abort + 180
5  llc                      0x0000000105e98e48 llvm::report_fatal_error(llvm::Twine const&, bool) + 328
6  llc                      0x0000000105e98d00 llvm::report_fatal_error(llvm::Twine const&, bool) + 0
7  llc                      0x0000000105a79690 llvm::MCAssembler::relaxBoundaryAlign(llvm::MCAsmLayout&, llvm::MCBoundaryAlignFragment&) + 0
8  llc                      0x0000000105a79d64 llvm::MCAssembler::layoutSectionOnce(llvm::MCAsmLayout&, llvm::MCSection&) + 72
9  llc                      0x0000000105a78ecc llvm::MCAssembler::layoutOnce(llvm::MCAsmLayout&) + 112
10 llc                      0x0000000105a78a88 llvm::MCAssembler::layout(llvm::MCAsmLayout&) + 404
11 llc                      0x0000000105a78ffc llvm::MCAssembler::Finish() + 64
12 llc                      0x0000000105a976a8 (anonymous namespace)::MCMachOStreamer::finishImpl() + 644
13 llc                      0x000000010554014c llvm::AsmPrinter::doFinalization(llvm::Module&) + 3232
14 llc                      0x00000001059fe330 llvm::FPPassManager::doFinalization(llvm::Module&) + 76
15 llc                      0x00000001059fa568 llvm::legacy::PassManagerImpl::run(llvm::Module&) + 960
16 llc                      0x0000000104d2d6e8 main + 6100
17 dyld                     0x00000001a8457f28 start + 2236
Abort trap: 6
```

rdar://113994760

This regression happened sometime between clang-15 and clang-16.  Bisection points at 648ce3d358560c0f60340fcf28ad2563d213eca2:
```
commit 648ce3d358560c0f60340fcf28ad2563d213eca2
Author: James Y Knight <jyknight@google.com>
Date:   Fri Jan 6 13:26:03 2023 -0500

    Cleanup unwind table emission code a bit.
    
    This change removes the `tidyLandingPads` function, which previously
    had a few responsibilities:
    
    1. Dealing with the deletion of an invoke, after MachineFunction lowering.
    2. Dealing with the deletion of a landing pad BB, after MachineFunction lowering.
    3. Cleaning up the type-id list generated by `MachineFunction::addLandingPad`.
    
    Case 3 has been fixed in the generator, and the others are now handled
    during table emission.
    
    This change also removes `MachineFunction`'s `addCatchTypeInfo`,
    `addFilterTypeInfo`, and `addCleanup` helper fns, as they had a single
    caller, and being outlined didn't make it simpler.
    
    Finally, as calling `tidyLandingPads` was effectively the only thing
    `DwarfCFIExceptionBase` did, that class has been eliminated.

```

### PR DESCRIPTION
**Pull request by jroelofs**
lowerInvokeable wasn't updating the returned chain after emitting the lowerEndEH, which caused SwiftErrorVal-handling code to re-set the DAG root, and thus accidentally skip the EH_LABEL node it was supposed to have addeed. After fixing that, a few places needed to be adjusted that assume the specific shape of the returned DAG.

Fixes: #64826
Fixes: rdar://113994760

### CODE REVIEW

File: llvm/test/CodeGen/X86/issue64826-switferror-eh.ll [hunk 1]
```diff
@@ -0,0 +1,75 @@
+; RUN: llc %s -filetype=obj -o - | llvm-readobj -r - | FileCheck %s --check-prefix=RELOC
+; RUN: llc %s -mtriple=x86_64-apple-macosx -o - | FileCheck %s --check-prefix=ASM
+
+target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-apple-macosx"
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/issue64826-switferror-eh.ll:5]: Either add the triple to both run lines and drop the IR one, or drop the triple from both command lines

File: llvm/test/CodeGen/X86/issue64826-switferror-eh.ll [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+; ASM-NEXT:    .uleb128 Ltmp1-Ltmp0                    ##   Call between Ltmp0 and Ltmp1
+; ASM-NEXT:    .uleb128 Ltmp2-Lfunc_begin0             ##     jumps to Ltmp2
+; ASM-NEXT:    .byte   0                               ##   On action: cleanup
+; ASM-NEXT:    .uleb128 Ltmp1-Lfunc_begin0             ## >> Call Site 2 <<
+; ASM-NEXT:    .uleb128 Lfunc_end0-Ltmp1               ##   Call between Ltmp1 and Lfunc_end0
+; ASM-NEXT:    .byte   0                               ##     has no landing pad
+; ASM-NEXT:    .byte   0                               ##   On action: cleanup
+; ASM-NEXT: Lcst_end0:
+; ASM-NEXT:    .p2align        2, 0x0
+; ASM-NEXT:                                         ## -- End function
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/issue64826-switferror-eh.ll:75]: missing newline end of file

File: llvm/test/CodeGen/X86/issue64826-switferror-eh.ll [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+; ASM-NEXT:  Lfunc_end0:
+; ASM-NEXT:    .cfi_endproc
+; ASM-NEXT:    .section        __TEXT,__gcc_except_tab
+; ASM-NEXT:    .p2align        2, 0x0
+; ASM-NEXT: GCC_except_table0:
+; ASM-NEXT: Lexception0:
+; ASM-NEXT:    .byte   255                             ## @LPStart Encoding = omit
+; ASM-NEXT:    .byte   255                             ## @TType Encoding = omit
+; ASM-NEXT:    .byte   1                               ## Call site Encoding = uleb128
+; ASM-NEXT:    .uleb128 Lcst_end0-Lcst_begin0
```
Comment (jroelofs) [inline llvm/test/CodeGen/X86/issue64826-switferror-eh.ll:63]: I would have used UTC, but it drops these :(

**Comment by jroelofs** [comment]
ping

File: llvm/test/CodeGen/X86/issue64826-switferror-eh.ll [hunk 4]
```diff
@@ -0,0 +1,75 @@
+; RUN: llc %s -filetype=obj -o - | llvm-readobj -r - | FileCheck %s --check-prefix=RELOC
+; RUN: llc %s -o - | FileCheck %s --check-prefix=ASM
+
+target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-apple-macosx"
+
+define swiftcc void @rdar113994760() personality ptr @__gcc_personality_v0 {
+entry:
+  %swifterror = alloca swifterror ptr, align 8
+  invoke swiftcc void null(i64 0, ptr null, ptr swifterror %swifterror)
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/issue64826-switferror-eh.ll:10]: avoiding the undefined call to null with something else would be slightly better

File: llvm/lib/CodeGen/SelectionDAG/StatepointLowering.cpp [hunk 5]
```diff
@@ -340,6 +340,9 @@ static std::pair<SDValue, SDNode *> lowerCallFromStatepointLoweringInfo(
   // to grab the return value from the return register(s), or it can be a LOAD
   // to load a value returned by reference via a stack slot.
 
+  if (CallEnd->getOpcode() == ISD::EH_LABEL)
+    CallEnd = CallEnd->getOperand(0).getNode();
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/StatepointLowering.cpp:344]: There's no statepoint being lowered in any test?

Comment (jroelofs) [reply llvm/lib/CodeGen/SelectionDAG/StatepointLowering.cpp:344]: These cover that path via the assertion on line 355:
```
 LLVM :: CodeGen/X86/statepoint-cmp-sunk-past-statepoint.ll
 LLVM :: CodeGen/X86/statepoint-invoke.ll
 LLVM :: CodeGen/X86/statepoint-ra.ll
 LLVM :: CodeGen/X86/statepoint-spill-lowering.ll
 LLVM :: CodeGen/X86/statepoint-stack-usage.ll
 LLVM :: CodeGen/X86/statepoint-vreg-invoke.ll
 ```

I'll re-run UTC and see what changes.

**Review by arsenm — APPROVED** [review]

**Pull request merged**